### PR TITLE
Add params to nb-tester

### PR DESCRIPTION
**Pull request by emilkovacev**
Migrating from #2344

Effectively allows the nb-tester script to be run against any backend. Also allows you to pass authentication and configuration directly to QiskitRuntimeService (if preferred).

Here's an example of how you can use the new changes to run a set of notebooks against test_eagle using qiskit_ibm_runtime:

```
test-docs-notebooks --config-path scripts/config/notebook-testing.toml --provider "qiskit_ibm_runtime" --backend "test_eagle2"
```

By default, `--provider` is set to `qiskit_fake_provider`, which runs the same provider code that was present in nb_tester before. By changing the provider, you can also specify "runtime_fake_provider" to use the fake_providers in `qiskit_ibm_runtime.fake_provider`, or "qiskit_ibm_runtime" to rely on real backends, and the credentials provided.

This shouldn't require changes in CI *yet* - it seems like notebook testing CI is able to complete the same number of tests as in main, although it does carry over some errors that I ran into on main too.

### CODE REVIEW

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:442]: Why use `_`? Usually CLIs use `-` as the delimiter

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:476]: Can you please add an example?

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:468]: Any helpful examples?

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:460]:
```suggestion
            'IBM Cloud API key or IBM Quantum API token. Warning: for security, you should set this via an environment variable, e.g. `--token="${IQP_API_TOKEN}"'
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:52]: Rather than using `kwargs`, which is vague/very generic, you can call this `runtime_service_args: dict[str, str | None]`. It will be more clear and keeps `**kwargs` available for future changes to this API.

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:50]:
```suggestion
    backend_name: str, 
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:53]:
```suggestion
) -> str:
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:59]: No need for a list comprehension - you can use a generator expression. That's more efficient

```suggestion
    qiskit_runtime_service_args = ", ".join(f"{arg}=\"{val}\"" if val else f"{arg}=None" for arg, val in kwargs.items())
```

Comment (Eric-Arellano) [reply scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:59]: Also this variable should only be defined in the block for qiskit_ibm_runtime because it's the only one that uses it

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:93]:
```suggestion
    patch += "\nQiskitRuntimeService.least_busy = patched_least_busy\n"
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:309]: This shouldn't be necessary? argparse will already handle for you converting `-` into `_` and applying defaults. It's a code smell that you have to specify the default twice atm.

What was the issue with usin `config.args.backend` et al?

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (emilkovacev) [reply scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:473]: For the examples, would it be more helpful to link the QiskitRuntimeService page and explicitly say to reference that page for those args, to use the descriptions from that page, or to come up with unique descriptions?

Comment (Eric-Arellano) [reply scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:473]: Linking to https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.QiskitRuntimeService makes sense. We want to make clear these are args for QiskitRuntimeService

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:62]: Maybe make `num_qubits` an option with default of 6 and as a new argument group called for qiskit-fake-provider?

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:432]:
```suggestion
            "`qiskit-ibm-runtime` or `runtime-fake-provider`."
```

**Review by Eric-Arellano — APPROVED** [review]
This looks great! Please wait for @frankharkins to approve before merging

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (frankharkins) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:254]: I noticed in CI it says we're patching with `fake_athens` even though we're patching with the Qiskit generic backend. I think easiest is to just remove the backend name. We can always check the command to see how we patched it.

```suggestion
        print(f"▶️ Executing {path} (with least_busy patched)")
```

**Review by frankharkins — COMMENTED** [review]
Looks great! Just one last thing

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (emilkovacev) [reply scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:83]: The only issue with this example (which I ran into while trying this out) is if the args aren't given, then config.args.channel is "None", which is invalid - which either means we write the following for each option:
```
f'channel="{config.args.channel}"' if config.args.channel else "channel=None"
```

Or, I could instead make that initial long join statement into a generic function since we would use it multiple times (and are likely to make use of it again):
```
def render_kwarg(arg, val):
    if isinstance(val, str):
        return f"{arg}=\"{val}\""
    return f"{arg}={val}"
```

and if there are any more use cases we can expand the usage of the function.

Comment (Eric-Arellano) [reply scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:83]: > render_kwarg

Sounds good to me.

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:70]:
```suggestion
        qiskit_fake_provider_args = render_kwargs(generic_backend_kwargs or {})
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:90]: We prefer self-documenting code over comments, unless when the comment adds something you can't easily infer

```suggestion
        qiskit_runtime_service_args = render_kwargs(runtime_service_kwargs or {})
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:451]: Nit to use whitespace to make the code more readable

```suggestion
    )

    generic_backend_options_group = parser.add_argument_group(
```

File: scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/__init__.py:477]:
```suggestion
    )

```

**Review by Eric-Arellano — APPROVED** [review]
Looks great! Feel free to merge after those last suggestions. Thanks for iterating on this!

**Pull request merged**
